[TOPIC-GPIO] Handle logical interrupt setting in common layer

galak opened the pull request:
Move handling of inverting of flags for gpio_pin_interrupt_configure to common code so drivers don't have to do it.

galak committed:
gpio: Handle logical conversion in gpio api layer

Move handling of logical flag support into gpio_pin_configure and
gpio_pin_interrupt_configure.  This way drivers don't need to know
anything about logical levels.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>
Signed-off-by: Piotr Mienkowski <piotr.mienkowski@gmail.com>

galak committed:
drivers: gpio_mcux: Update for handling local layer in core

Remove handling for GPIO_INT_LEVELS_LOGICAL in driver now that we do
that in gpio_pin_interrupt_configure and gpio_pin_configure.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
drivers: gpio_gecko: Update for handling local layer in core

Remove handling for GPIO_INT_LEVELS_LOGICAL in driver now that we do
that in gpio_pin_interrupt_configure and gpio_pin_configure.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
drivers: gpio_sam: Update for handling local layer in core

Remove handling for GPIO_INT_LEVELS_LOGICAL in driver now that we do
that in gpio_pin_interrupt_configure and gpio_pin_configure.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
drivers: gpio_nrfx: Update for handling local layer in core

Remove handling for GPIO_INT_LEVELS_LOGICAL in driver now that we do
that in gpio_pin_interrupt_configure and gpio_pin_configure.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

mnkp commented:
Handling of interrupt logical levels directly in the `include/drivers/gpio.h` does simplify the driver code, makes it more readable and robust. I would do it a bit differently though. Rather than replacing the bit fields in the flags parameter, which has a potential to confuse some users, I would translate the bit fields to a new, clean API which is internal to the driver. I.e. we could define the `pin_interrupt_configure` function as:
```
enum gpio_int_mode {
	GPIO_INT_MODE_DISABLED,
	GPIO_INT_MODE_LEVEL,
	GPIO_INT_MODE_EDGE,
};

enum gpio_int_trigger {
	GPIO_INT_TRIGGER_LOW = 1,
	GPIO_INT_TRIGGER_HIGH,
	GPIO_INT_TRIGGER_BOTH,
};

int (*pin_interrupt_configure)(struct device *port, unsigned int pin,
			       enum gpio_int_mode mode, enum gpio_int_trigger trigger);
```
This part would not be visible in the public API, i.e. it wouldn't have public doxygen documentation. 

This approach has also potential of making #19251 redundant.

galak commented:
> ```
> enum gpio_int_mode {
> 	GPIO_INT_MODE_DISABLED,
> 	GPIO_INT_MODE_LEVEL,
> 	GPIO_INT_MODE_EDGE,
> };
> 
> enum gpio_int_trigger {
> 	/* Trigger detection when input state is (or transitions to) physical low. */
> 	GPIO_INT_TRIGGER_LOW = 1,
> 	/* Trigger detection when input state is (or transitions to) physical high. */
> 	GPIO_INT_TRIGGER_HIGH,
> 	/* Trigger detection on pin rising or falling edge. */
> 	GPIO_INT_TRIGGER_BOTH,
> };
> 
> int (*pin_interrupt_configure)(struct device *port, unsigned int pin,
> 			       enum gpio_int_mode mode, enum gpio_int_trigger trigger);
> ```

If we go done this path I would suggest a single enum rather than having to deal with invalid combinations of gpio_int_mode & gpio_int_trigger.

Unfortunately this would mean a conversion between the existing flags and `gpio_int_trigger_mode`

mnkp commented:
> If we go done this path I would suggest a single enum rather than having to deal with invalid combinations of gpio_int_mode & gpio_int_trigger.

I'm not sure about this one. We would need then to decode the enum in the driver. That's going to be a lot of if statements (we shouldn't treat enum as a flag value anymore). The driver code has only a few decision it needs to take: should I enable/disable interrupts. If I should enable interrupts should it be level or edge. If it's edge or level, should it be falling/low, rising/high or both. In the hardware information about falling/low and rising/high trigger is often configured using the same register bits. The two argument approach will result in a smaller code size. Also, the only invalid combination is level interrupt and trigger on both edges. If we are concern about scenario `mode == GPIO_INT_MODE_DISABLED` and `trigger == GPIO_INT_TRIGGER_HIGH` I would rather cover this case via an extra ASSERT in the `gpio_pin_interrupt_configure()` function.

galak committed:
gpio: Reorder code in drivers/gpio.h

Move z_impl_gpio_pin_interrupt_configure before gpio_pin_configure so we
can utilize z_impl_gpio_pin_interrupt_configure in gpio_pin_configure in
the future.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
drivers: gpio_altera_nios2: Add gpio_driver_data to driver data

Add gpio_driver_data as the first element in the driver data as the gpio
core expects this.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
drivers: gpio_cc13xx_cc26xx: Add gpio_driver_data to driver data

Add gpio_driver_data as the first element in the driver data as the gpio
core expects this.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
drivers: gpio_cc2650: Add gpio_driver_data to driver data

Add gpio_driver_data as the first element in the driver data as the gpio
core expects this.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
drivers: gpio_cc32xx: Add gpio_driver_data to driver data

Add gpio_driver_data as the first element in the driver data as the gpio
core expects this.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
drivers: gpio_cmsdk_ahb: Add gpio_driver_data to driver data

Add gpio_driver_data as the first element in the driver data as the gpio
core expects this.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
drivers: gpio_dw: Add gpio_driver_data to driver data

Add gpio_driver_data as the first element in the driver data as the gpio
core expects this.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
drivers: gpio_esp32: Add gpio_driver_data to driver data

Add gpio_driver_data as the first element in the driver data as the gpio
core expects this.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
drivers: gpio_ht16k33: Add gpio_driver_data to driver data

Add gpio_driver_data as the first element in the driver data as the gpio
core expects this.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
drivers: gpio_imx: Add gpio_driver_data to driver data

Add gpio_driver_data as the first element in the driver data as the gpio
core expects this.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
drivers: gpio_intel_apl: Add gpio_driver_data to driver data

Add gpio_driver_data as the first element in the driver data as the gpio
core expects this.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
drivers: gpio_mchp_xec: Add gpio_driver_data to driver data

Add gpio_driver_data as the first element in the driver data as the gpio
core expects this.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
drivers: gpio_mcux_lpc: Add gpio_driver_data to driver data

Add gpio_driver_data as the first element in the driver data as the gpio
core expects this.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
drivers: gpio_pcal9535a: Add gpio_driver_data to driver data

Add gpio_driver_data as the first element in the driver data as the gpio
core expects this.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
drivers: gpio_rv32m1: Add gpio_driver_data to driver data

Add gpio_driver_data as the first element in the driver data as the gpio
core expects this.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
drivers: gpio_sam0: Add gpio_driver_data to driver data

Add gpio_driver_data as the first element in the driver data as the gpio
core expects this.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
drivers: gpio_sifive: Add gpio_driver_data to driver data

Add gpio_driver_data as the first element in the driver data as the gpio
core expects this.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
drivers: gpio_stellaris: Add gpio_driver_data to driver data

Add gpio_driver_data as the first element in the driver data as the gpio
core expects this.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
drivers: gpio_stm32: Add gpio_driver_data to driver data

Add gpio_driver_data as the first element in the driver data as the gpio
core expects this.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
drivers: gpio_sx1509b: Add gpio_driver_data to driver data

Add gpio_driver_data as the first element in the driver data as the gpio
core expects this.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
drivers: gpio_mmio32: Add gpio_driver_data to driver data

Add gpio_driver_data as the first element in the driver data as the gpio
core expects this.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak commented:
I didn't update `gpio_mcux_igpio.c` as @MaureenHelm already has a PR outstanding for that.

pabigot commented on include/drivers/gpio.h:
```diff
@@ -542,6 +564,64 @@ static inline int z_impl_gpio_disable_callback(struct device *port,
+	flags &= ~GPIO_INT_LEVELS_LOGICAL;
+
+	if ((flags == GPIO_INT_MODE_DISABLED) ||
+	    (flags == GPIO_INT_LEVEL_LOW) ||
```
Shouldn't this be `GPIO_INT_MODE_LEVEL_LOW`?

galak replied:
Yes, good catch.

galak replied:
Fixed.

mnkp commented on include/drivers/gpio.h:
```diff
@@ -568,14 +589,37 @@ static inline int z_impl_gpio_pin_interrupt_configure(struct device *port,
+	}
+
+	return -EINVAL;
 }
```
I didn't analyse this code, however it doesn't pass the `tests/drivers/gpio/gpio_api_1pin` testcase. Instead the function can be written as
```
static inline int z_impl_gpio_pin_interrupt_configure(struct device *port,
		unsigned int pin, unsigned int flags)
{
	const struct gpio_driver_api *api =
		(const struct gpio_driver_api *)port->driver_api;
	struct gpio_driver_data *const data =
		(struct gpio_driver_data *const)port->driver_data;

	__ASSERT(pin < GPIO_MAX_PINS_PER_PORT, "Invalid pin number");

	__ASSERT(((flags & GPIO_INT_ENABLE) == 0) ||
		 ((flags & (GPIO_INT_LOW_0 | GPIO_INT_HIGH_1)) != 0),
		 "At least one of GPIO_INT_LOW_0, GPIO_INT_HIGH_1 has to be "
		 "enabled.");

	if (((flags & GPIO_INT_LEVELS_LOGICAL) != 0) &&
	    ((data->invert & BIT(pin)) != 0)) {
		/* Invert signal bits */
		flags ^= (GPIO_INT_LOW_0 | GPIO_INT_HIGH_1);
	}

	flags &= (GPIO_INT_LOW_0 | GPIO_INT_HIGH_1 | GPIO_INT_EDGE |
		  GPIO_INT_ENABLE);

	return api->pin_interrupt_configure(port, pin,
			(enum gpio_int_trigger_mode)flags);
}
```

mnkp commented on include/drivers/gpio.h:
```diff
@@ -451,6 +457,21 @@ struct gpio_callback {
+	GPIO_INT_MODE_EDGE_RISING = GPIO_INT_EDGE | GPIO_INT_HIGH_1,
+	GPIO_INT_MODE_EDGE_BOTH = GPIO_INT_EDGE | GPIO_INT_LOW_0 |
+				  GPIO_INT_HIGH_1,
+};
```
This can be implemented simpler as
```
	GPIO_INT_MODE_DISABLED = GPIO_INT_DISABLE,
	GPIO_INT_MODE_LEVEL_LOW = GPIO_INT_LEVEL_LOW,
	GPIO_INT_MODE_LEVEL_HIGH = GPIO_INT_LEVEL_HIGH,
	GPIO_INT_MODE_EDGE_FALLING = GPIO_INT_EDGE_FALLING,
	GPIO_INT_MODE_EDGE_RISING = GPIO_INT_EDGE_RISING,
	GPIO_INT_MODE_EDGE_BOTH = GPIO_INT_EDGE_BOTH,
```

mnkp commented on include/drivers/gpio.h:
```diff
@@ -451,6 +457,21 @@ struct gpio_callback {
+ * in terms of the public int flags so we can just mask and pass them
+ * through to the driver api
+ */
+enum gpio_int_trigger_mode {
```
The enum name should match the name of its values, i.e. we should either rename the enum to `enum gpio_int_mode` or use `GPIO_INT_TRIGGER_MODE_` as a prefix for the enum values.

galak replied:
Changed enum name.

mnkp commented on include/drivers/gpio.h:
```diff
@@ -112,6 +112,12 @@ extern "C" {
 #define GPIO_INT_LEVELS_LOGICAL        (1U << 16)
 
+/** @cond INTERNAL_HIDDEN */
+#define GPIO_INT_ALL_MASK (GPIO_INT_LOW_0 | GPIO_INT_HIGH_1 | \
```
Not required by the new implementation of `gpio_pin_interrupt_configure` function.

mnkp reviewed (changes requested):
The current implementation doesn't pass the `tests/drivers/gpio/gpio_api_1pin` testcase. I provided alternative implementation in the comments. After applying all suggestions the testcase should pass. I run it on sam_e70_xplained, nrf52840_pca10056, efr32_slwstk6061a boards. Unfortunately, the gpio_mcux driver can't be tested with the 1 pin testcase.

I propose to drop "gpio: Renumber IRQ flag bit positions to make code more efficient" commit. It's not necessary.

mnkp commented on include/drivers/gpio.h:
```diff
@@ -602,7 +652,21 @@ static inline int gpio_pin_configure(struct device *port, u32_t pin,
+			ret = z_impl_gpio_pin_interrupt_configure(port, pin,
+								  flags);
+		}
+	}
```
To avoid nesting this part could be written as
```
	ret = gpio_config(port, GPIO_ACCESS_BY_PIN, pin, flags);
	if (ret != 0) {
		return ret;
	}

	if ((flags & GPIO_ACTIVE_LOW) != 0) {
		data->invert |= BIT(pin);
	} else {
		data->invert &= ~BIT(pin);
	}
	if (api->pin_interrupt_configure) {
		ret = z_impl_gpio_pin_interrupt_configure(port, pin, flags);
	}
```

mnkp commented on include/drivers/gpio.h:
```diff
@@ -594,6 +638,12 @@ static inline int z_impl_gpio_pin_interrupt_configure(struct device *port,
+	const struct gpio_driver_api *api =
+		(const struct gpio_driver_api *)port->driver_api;
+	struct gpio_driver_data *data =
+			(struct gpio_driver_data *)port->driver_data;
```
alignment

mnkp commented on drivers/gpio/gpio_nrfx.c:
```diff
@@ -328,19 +316,36 @@ static int gpio_nrfx_pin_interrupt_configure(struct device *port,
+	case GPIO_INT_MODE_EDGE_BOTH:
+		edge = true;
+		WRITE_BIT(data->double_edge, pin, true);
+		break;
```
That's what I had in mind when I proposed 
```
int (*pin_interrupt_configure)(struct device *port, unsigned int pin,
			       enum gpio_int_mode mode, enum gpio_int_trigger trigger);
```
as a signature to the configure function. There would be no need to decode the meaning of the enum in the driver. Or at least the decoding process would be simpler. The above `switch` and the following `WRITE_BIT` macro translates into a lot of if statements.

mnkp commented on drivers/gpio/gpio_nrfx.c:
```diff
@@ -311,13 +298,14 @@ static int gpio_nrfx_port_toggle_bits(struct device *port, u32_t mask)
-	    ((flags & GPIO_INT_EDGE) != 0) &&
+	if (((flags == GPIO_INT_MODE_EDGE_FALLING) ||
+	     (flags == GPIO_INT_MODE_EDGE_RISING)  ||
+	     (flags == GPIO_INT_MODE_EDGE_BOTH)) &&
```
This long if condition would be much simpler if we split `enum gpio_int_trigger_mode` into `enum gpio_int_mode mode, enum gpio_int_trigger trigger`.

mnkp commented on drivers/gpio/gpio_nrfx.c:
```diff
@@ -328,19 +316,36 @@ static int gpio_nrfx_pin_interrupt_configure(struct device *port,
+		break;
+	case GPIO_INT_MODE_EDGE_BOTH:
+		edge = true;
+		WRITE_BIT(data->double_edge, pin, true);
```
The `gpio_api_1pin_test` didn't catch it but this looks like an error. If we run `gpio_pin_interrupt_configure` to enable interrupt on both edges and then re run the function on the same pin to enable interrupt on the rising edge only the `data->double_edge` will be stuck at `1` indicating double edge interrupt.

pabigot commented on include/drivers/gpio.h:
```diff
@@ -566,7 +652,21 @@ static inline int gpio_pin_configure(struct device *port, u32_t pin,
-	return gpio_config(port, GPIO_ACCESS_BY_PIN, pin, flags);
+	ret = gpio_config(port, GPIO_ACCESS_BY_PIN, pin, flags);
+
+	if (ret == 0) {
```
This block seems to change the API contract.  In the merged design (and still documented) `gpio_pin_configure()` does not interpret interrupt configuration.  Thus pin configuration and interrupt configuration are completely separated.

With this implementation configuring a pin appears to also configure the interrupts.  That seems to be mixing capabilities that were intended (as I understood it) to be kept separate in the new API.

mnkp replied:
This is required to keep backward compatibility with the existing application code. Will be removed in the future. As I have mentioned already, I'm not really sure how to handle deprecation of a feature like this one so suggestions are welcome.

galak commented:
> The current implementation doesn't pass the `tests/drivers/gpio/gpio_api_1pin` testcase. I provided alternative implementation in the comments. After applying all suggestions the testcase should pass. I run it on sam_e70_xplained, nrf52840_pca10056, efr32_slwstk6061a boards. Unfortunately, the gpio_mcux driver can't be tested with the 1 pin testcase.

curious why the 1 pin testcase doesn't work on mcux?

mnkp commented:
> curious why the 1 pin testcase doesn't work on mcux?

mcux driver doesn't allow to configure pins in in/out mode. It's a hardware limitation. The pin can be either input or output but not both. In this case the 1 pin testcase still needs to pass but pretty much all the test cases will be skipped.

galak commented:
> mcux driver doesn't allow to configure pins in in/out mode. It's a hardware limitation. The pin can be either input or output but not both. In this case the 1 pin testcase still needs to pass but pretty much all the test cases will be skipped.

Ah with the other platforms you get a kinda internal lookback on a single pin.  With mcux we need 2 pins with a physical wire.

mnkp commented:
> Ah with the other platforms you get a kinda internal lookback on a single pin. With mcux we need 2 pins with a physical wire.

Yes, that's why it would be good to have the 2 pin testcase working, on the topic branch.

mnkp commented on drivers/gpio/gpio_sam.c:
```diff
@@ -256,48 +245,44 @@ static int gpio_sam_port_toggle_bits(struct device *dev, u32_t mask)
 		pio->PIO_AIMER = mask;
 
-		if (flags & GPIO_INT_EDGE) {
+		if (mode == GPIO_INT_EDGE) {
```
`GPIO_INT_MODE_EDGE`

galak replied:
Done.

mnkp commented on drivers/gpio/gpio_nrfx.c:
```diff
@@ -328,19 +315,26 @@ static int gpio_nrfx_pin_interrupt_configure(struct device *port,
+	WRITE_BIT(data->pin_int_en, pin, pin_int_en);
+	WRITE_BIT(data->int_en, pin, true);
+	WRITE_BIT(data->trig_edge, pin, edge);
+	WRITE_BIT(data->int_active_level, pin, active);
```
Doesn't pass the 1 pin testcase. Following version does:
```
static int gpio_nrfx_pin_interrupt_configure(struct device *port,
		unsigned int pin, enum gpio_int_mode mode,
	       	enum gpio_int_trig trig)
{
	struct gpio_nrfx_data *data = get_port_data(port);
	u32_t abs_pin = NRF_GPIO_PIN_MAP(get_port_cfg(port)->port_num, pin);

	if ((mode == GPIO_INT_MODE_EDGE) &&
	    (nrf_gpio_pin_dir_get(abs_pin) == NRF_GPIO_PIN_DIR_OUTPUT)) {
		/*
		 * The pin's output value as specified in the GPIO will be
		 * ignored as long as the pin is controlled by GPIOTE.
		 * Pin with output enabled cannot be used as an edge interrupt
		 * source.
		 */
		return -ENOTSUP;
	}

	WRITE_BIT(data->pin_int_en, pin, mode != GPIO_INT_MODE_DISABLED);
	WRITE_BIT(data->int_en, pin, mode != GPIO_INT_MODE_DISABLED);
	WRITE_BIT(data->trig_edge, pin, mode == GPIO_INT_MODE_EDGE);
	WRITE_BIT(data->double_edge, pin, trig == GPIO_INT_TRIG_BOTH);
	WRITE_BIT(data->int_active_level, pin, trig == GPIO_INT_TRIG_HIGH);

	return gpiote_pin_int_cfg(port, pin);
}
```

galak replied:
Done.

mnkp commented on drivers/gpio/gpio_gecko.c:
```diff
@@ -286,22 +270,27 @@ static int gpio_gecko_pin_interrupt_configure(struct device *dev,
+		en = false;
+	}
+
+	WRITE_BIT(data->pin_callback_enables, pin, en);
```
This works well but can be written simpler as
```
	WRITE_BIT(data->pin_callback_enables, pin, mode != GPIO_INT_DISABLE);

```

galak replied:
Done.

mnkp commented on drivers/gpio/gpio_mcux.c:
```diff
@@ -288,16 +276,21 @@ static int gpio_mcux_pin_interrupt_configure(struct device *dev,
+		en = false;
+	}
+
+	WRITE_BIT(data->pin_callback_enables, pin, en);
```
Can be written as
```
	WRITE_BIT(data->pin_callback_enables, pin, mode != GPIO_INT_DISABLE);
```

galak replied:
Done.

mnkp commented on include/drivers/gpio.h:
```diff
@@ -450,6 +450,23 @@ struct gpio_callback {
+enum gpio_int_trig {
+	GPIO_INT_TRIG_LOW = GPIO_INT_LOW_0,
+	GPIO_INT_TRIG_HIGH = GPIO_INT_HIGH_1,
+	GPIO_INT_TRIG_BOTH = GPIO_INT_LOW_0 | GPIO_INT_HIGH_1,
```
Let's add description of the trigger modes since one flag covers both a state and a transition.
```
	/* Trigger detection when input state is (or transitions to) physical low. */
	GPIO_INT_TRIG_LOW = GPIO_INT_LOW_0,
	/* Trigger detection when input state is (or transitions to) physical high. */
	GPIO_INT_TRIG_HIGH = GPIO_INT_HIGH_1,
	/* Trigger detection on pin rising or falling edge. */
	GPIO_INT_TRIG_BOTH = GPIO_INT_LOW_0 | GPIO_INT_HIGH_1,
```

galak replied:
Updated.

mnkp reviewed (approved):
Tested with `tests/drivers/gpio/gpio_api_1pin` on sam_e70_xplained, efr32_slwstk6061a, nrf52840_pca10056 evaluation boards and running button sample app on frdm_k64f board.

mnkp commented:
One minor comment, maybe we should squash all the different "drivers: *: Add gpio_driver_data to driver data" commits into a single one? Logically it is a single change.

pabigot commented:
> One minor comment, maybe we should squash all the different "drivers: *: Add gpio_driver_data to driver data" commits into a single one? Logically it is a single change.

That sounds good.  If that's done I'd also like to see the per-driver declaration switch to
```
    /* gpio_driver_data needs to be first */
    struct gpio_driver_data common;
```
and some documentation around:

https://github.com/zephyrproject-rtos/zephyr/blob/4bc09ff3e3d8e9b9e395a7d2bdd7cb2752e9453c/include/drivers/gpio.h#L397

to note that this structure is common to all drivers and must be placed first within the driver-specific `struct driver_data` declaration.  "general" doesn't convey the intent quite as well as "common" would, IMO.

mnkp commented:
> I'd also like to see the per-driver declaration switch to
> 
> ```
>     /* gpio_driver_data needs to be first */
>     struct gpio_driver_data common;
> ```

I was thinking about this but I feared that `common` would suggest that this piece of data is somehow shared between different driver instances, run time, which is not the case. I admit `general` is not the best name but `common` is not really better.

galak commented:
> One minor comment, maybe we should squash all the different "drivers: *: Add gpio_driver_data to driver data" commits into a single one? Logically it is a single change.
>

I want to leave them so we could squash each driver change into its own commit before sending the branch upstream.
 
> We need one more update and Shippable run since the checkpatch is still failing. Looks like a few legitimate issues.

will take a look.

galak commented:
> > One minor comment, maybe we should squash all the different "drivers: *: Add gpio_driver_data to driver data" commits into a single one? Logically it is a single change.
> 
> That sounds good. If that's done I'd also like to see the per-driver declaration switch to
> 
> ```
>     /* gpio_driver_data needs to be first */
>     struct gpio_driver_data common;
> ```
> 
> and some documentation around:
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/4bc09ff3e3d8e9b9e395a7d2bdd7cb2752e9453c/include/drivers/gpio.h#L397
> 
> to note that this structure is common to all drivers and must be placed first within the driver-specific `struct driver_data` declaration. "general" doesn't convey the intent quite as well as "common" would, IMO.

Added comment about `struct gpio_driver_data`.

galak commented:
> I was thinking about this but I feared that `common` would suggest that this piece of data is somehow shared between different driver instances, run time, which is not the case. I admit `general` is not the best name but `common` has its own issues.

`drv_data`, `gpio_drv_dat`, `priv_data`??

Just let me know what you guys want it to be.
